src/main/clojure/clara/tools/fact_graph.clj

```diff
@@ -0,0 +1,96 @@
+(ns clara.tools.fact-graph
+  (:require [clara.tools.inspect :as i]
+            [schema.core :as sc]))
+
+;; This node will have either facts or results of accumulations as its parents.
+;; Its children will be facts that the rule inserted.
+(sc/defrecord RuleActivationNode [rule-name :- sc/Str
+                                  id :- sc/Int])
+
+;; The parents of this node are facts over which an accumulation was run.
+;; It will have a single child, the result of the accumulation.  So, for example,
+;; with the condition [?t <- (acc/min :temperature) :from [Temperature]], if we have
+;; (->Temperature 50 "MCI") and (->Temperature 30 "MCI") the child of this node will be
+;; an AccumulationResult with the :result 30 and the parents will be the two Temperature facts.
+(sc/defrecord AccumulationNode [id :- sc/Int])
+
+;; As alluded to above, this node represents the result of an accumulation.  Its child will be a
+;; RuleActivationNode.  Note that there will be an AccumulationResult for each distinct rules firing.
+(sc/defrecord AccumulationResultNode [id :- sc/Int
+                                      result :- sc/Any])
+
+(def ^:private empty-fact-graph {:forward-edges {}
+                                 :backward-edges {}})
+
+(defn ^:private add-edge [graph to from]
```

Comment:
It looks like I reversed the names on "to" and "from" but used them correctly so the tests still passed; will fix.